Fix max_transaction_time

javsanpar opened the pull request:
Fixes #3957

javsanpar committed:
Fix max_transaction_time boundaries between transactions

Fixes issue #3957

javsanpar committed:
Add test for max_transaction_time

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -4968,6 +4962,15 @@ int MySQL_Session::handler() {
 
+					if (active_transactions != 0) {  // run this only if currently we think there is a transaction
+						// Redundant?
+						//if ((myconn->mysql->server_status & SERVER_STATUS_IN_TRANS) == 0) { // there is no transaction on the backend connection
```
This is not redundant.
If `SERVER_STATUS_IN_TRANS` is set in `myconn->mysql->server_status` , there is an active transaction

javsanpar replied:
`bool MySQL_Connection::IsActiveTransaction()` does the same and `unsigned int MySQL_Session::NumActiveTransactions(bool check_savepoint)` calls it for all the backends of the session

renecannao replied:
I didn't understand. What does that mean?

javsanpar replied:
I mean that we do not need to do `if ((myconn->mysql->server_status & SERVER_STATUS_IN_TRANS) == 0)` because `MySQL_Session::NumActiveTransactions` already does it for the connection of each backend associated to the session.

renecannao replied:
And I mean that I think we don't need to call `NumActiveTransactions()` if `active_transaction != 0` and `(myconn->mysql->server_status & SERVER_STATUS_IN_TRANS) == 0)` .
Thoughts?

javsanpar replied:
You are right. So the objective of this code is to save unneeded fetching of `server_status` of the other connections if we know `myconn->mysql->server_status` is already in a transaction?

renecannao replied:
Right

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -5075,6 +5078,12 @@ int MySQL_Session::handler() {
 							handler_minus1_HandleBackendConnection(myds, myconn);
 						}
 					} else {
+						if (active_transactions == 0) {
```
Why not setting this at the *beginning* of a query?

javsanpar replied:
I think this is the beginning of a query. When `rc==0` is the end of the query and when `rc==-1` the query do not begin because it has failed.

What I am realizing now is that if `rc==-1` because proxysql failed to retrieve the result of a query, then the number of active transaction is not going to be updated and the `transaction_started_at` is not going to be reset.

javsanpar committed:
save unneeded fetching of all 'server_status'

javsanpar commented on lib/MySQL_Session.cpp:
```diff
@@ -4968,6 +4962,14 @@ int MySQL_Session::handler() {
+								transaction_started_at = 0; // reset it
+						}
+					}
+
```
@renecannao what do you think about putting this code inside a function and calling it here and also when `rc==-1`. This way, ProxySQL also reset the timer when it has an error in the middle of a transaction.

renecannao replied:
Why do you want to reset the timer if there is an error in the middle of a transaction?

I am going back to what I wrote before: [...] at the beginning of the query

javsanpar committed:
Set active_transactions and transaction_started_at when the query completes

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -4968,6 +4968,9 @@ int MySQL_Session::handler() {
 								transaction_started_at = 0; // reset it
 						}
+					} else {
+						transaction_started_at = thread->curtime;
```
This code means that when a query completes, and `active_transactions==0` , `transaction_started_at` is set to current time even if there is no transaction.

renecannao commented on test/tap/tests/test_max_transaction_time-t.cpp:
```diff
@@ -0,0 +1,67 @@
+		return EXIT_FAILURE;
+	}
+
+	diag("Configure the target server (non-existing) server to test connection failures");
```
This seems incorrect

javsanpar replied:
Right, solved

renecannao commented on test/tap/tests/test_max_transaction_time-t.cpp:
```diff
@@ -0,0 +1,67 @@
+		diag("Failed to get the required environmental variables.");
+		return EXIT_FAILURE;
+	}
+
```
`plan()` is missing

javsanpar replied:
Sorry, I thought it was not needed when not using `ok()`

renecannao committed:
Automatically start a transaction when running a query

This commit also introduces 2 output for PROXYSQL INTERNAL SESSION:
- active_transactions
- transaction_started_at

Enhanced the TAP tests to also verify the above metrics

javsanpar committed:
Fix informative message in test_max_transaction_time-t

renecannao committed:
More testing in test_max_transaction_time-t

IN PROXYSQL SESSION INTERNAL replace transaction_started_at with transaction_time_ms

renecannao committed:
Improving TAP test test_max_transaction_time-t

matus-chordify commented:
Great, thank you @javsanpar and @renecannao! :tada:
